src/Featurizers/TfidfVectorizerFeaturizer.h

```diff
@@ -0,0 +1,523 @@
+
+    // MSVC has problems when the definition and declaration are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        //termfrequency for specific document
+        MapWithIterRange documentTermFrequency;
+
+        //todo: will use vector<functor> after string header file is done
+        Strings::Parse<std::string::const_iterator>(
+            input,
+            [](char c) {return std::isspace(c);},
+            [&documentTermFrequency] (std::string::const_iterator iterStart, std::string::const_iterator iterEnd) {
+                MapWithIterRange::iterator docuTermFreqIter(documentTermFrequency.find(std::make_tuple(iterStart, iterEnd)));
+                if (docuTermFreqIter != documentTermFrequency.end()) {
+                    ++docuTermFreqIter->second;
+                } else {
+                    documentTermFrequency.insert(std::make_pair(std::make_tuple(iterStart, iterEnd), 1));
+                }
+            }
+        );
+
+        std::float_t normVal = 0.0f;
+        std::vector<std::tuple<std::uint32_t, std::float_t>> results;
+        for (auto const & pair : documentTermFrequency) {
+            std::string const word(std::string(std::get<0>(pair.first), std::get<1>(pair.first)));
+
+            IndexMap::const_iterator const      labelIter(Labels.find(word));
+
+            if (labelIter != Labels.end()) {
+
+                std::float_t tf;
+                std::float_t idf;
+
+                //calculate tf
+                if (Binary) {
+                    tf = 1.0f;
+                } else if (!SublinearTf) {
+                    tf = static_cast<std::float_t>(pair.second);
+                } else {
+                    tf = 1.0f + static_cast<std::float_t>(log(pair.second));
+                }
+
+                //calculate idf;
+                if (!UseIdf) {
+                    idf = 1.0f;
+                } else if (SmoothIdf) {
+                    idf = 1.0f + log((1 + TotalNumsDocuments) / (1.0f + static_cast<std::float_t>(DocumentFreq.at(word))));
+                } else {
+                    idf = 1.0f + log((1 + TotalNumsDocuments) / (static_cast<std::float_t>(DocumentFreq.at(word))));
+                }
+
+                //calculate tfidf
+                std::float_t tfidf = tf * idf;
+
+                //calculate normVal
+                if(Norm == "l1") {
+                    normVal += abs(tfidf);
+                } else if (Norm == "l2") {
+                    normVal += tfidf * tfidf;
+                }
+
+                //temperarily put output in a vector for future normalization
+                results.emplace_back(std::make_tuple(labelIter->second, tfidf));
+            }
+        }
+        //normVal will never be 0 as long as results is not empty
+        if (normVal == 0.0f)
+            throw std::runtime_error("This happens when the input document is empty");
```

Comment:
It seems like it would be good to explicitly check for an empty document either at the beginning of this method or right after its contents are parsed. With that check in place, we could assert that `normVal != 0.0f` here.